File: ScaleformUI_Lua/src/Menus/PauseMenus/PauseMenu/Tabs/TabLeftItem.lua
TabLeftItem = setmetatable({}, TabLeftItem)
TabLeftItem.__index = TabLeftItem
TabLeftItem.__call = function()
    return "TabLeftItem", "TabLeftItem"
end

---@class TabLeftItem
---@field public Label string
---@field public ItemType number
---@field public Focused boolean
---@field public MainColor SColor
---@field public HighlightColor SColor
---@field public Highlighted boolean
---@field public ItemIndex number
---@field public ItemList table<SettingsItem|SettingsListItem|SettingsProgressItem|SettingsCheckboxItem|SettingsSliderItem|KeymapItem>
---@field public RightTitle string
---@field private _enabled boolean
---@field private _hovered boolean
---@field private _selected boolean
---@field public KeymapRightLabel_1 string
---@field public KeymapRightLabel_2 string
---@field public Index number
---@field public AddItem fun(self:TabLeftItem, item:SettingsItem|SettingsListItem|SettingsProgressItem|SettingsCheckboxItem|SettingsSliderItem|KeymapItem|BasicTabItem)
---@field public Enabled fun(self:TabLeftItem, enabled:boolean):boolean
---@field public Hovered fun(self:TabLeftItem, hover:boolean):boolean
---@field public Selected fun(self:TabLeftItem, selected:boolean):boolean
---@field public OnIndexChanged fun(item:TabLeftItem, index:number)
---@field public OnActivated fun(item:TabLeftItem, index:number)

function TabLeftItem.New(label, _type, mainColor, highlightColor, labelFont)
    local __formatLeftLabel = (tostring(label))
    if not __formatLeftLabel:StartsWith("~") then
        __formatLeftLabel = "~s~" .. __formatLeftLabel
    end
    local data = {
        Label = label or "",
        ItemType = _type,
        Focused = false,
        MainColor = mainColor or SColor.HUD_Pause_bg,
        HighlightColor = highlightColor or SColor.HUD_White,
        Highlighted = false,
        ItemIndex = 0,
        ItemList = {} --[[@type table<SettingsItem|SettingsListItem|SettingsProgressItem|SettingsCheckboxItem|SettingsSliderItem|KeymapItem|BasicTabItem>]],
        RightTitle = "",
        _enabled = true,
        _hovered = false,
        _selected = false,
        _formatLeftLabel = __formatLeftLabel or "",
        KeymapRightLabel_1 = "",
        KeymapRightLabel_2 = "",
        TextureDict = "",
        TextureName = "",
        LeftItemBGType = 0,
        _labelFont = labelFont or ScaleformFonts.CHALET_LONDON_NINETEENSIXTY,
        --_rightLabelFont = ScaleformFonts.CHALET_LONDON_NINETEENSIXTY,
        OnIndexChanged = function(item, index)
        end,
        OnActivated = function(item, index)
        end,
        Index = 0,
        Parent = nil
    }
    return setmetatable(data, TabLeftItem)
end

---Add item to the tab list
---@param item SettingsItem|SettingsListItem|SettingsProgressItem|SettingsCheckboxItem|SettingsSliderItem|KeymapItem|BasicTabItem
function TabLeftItem:AddItem(item)
    item.Parent = self
    self.ItemList[#self.ItemList + 1] = item
end

function TabLeftItem:Label(label)
    if label ~= nil then
        self.Label = label
        self._formatLeftLabel = tostring(Text)
        if not self._formatLeftLabel:StartsWith("~") then
            self._formatLeftLabel = "~s~" .. self._formatLeftLabel
        end
        if self:Selected() then
            self._formatLeftLabel = self._formatLeftLabel:gsub("~w~", "~l~")
            self._formatLeftLabel = self._formatLeftLabel:gsub("~s~", "~l~")
        else
            self._formatLeftLabel = self._formatLeftLabel:gsub("~l~", "~s~")
        end
        if self.Parent ~= nil and self.Parent.Base.Parent ~= nil and self.Parent.Base.Parent:Visible() then
            local tab = IndexOf(self.Parent.Base.Parent.Tabs, self.Parent) - 1
            local leftItem = IndexOf(self.Parent.LeftItemList, self) - 1
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("UPDATE_LEFT_ITEM_LABEL", tab, leftItem, self._formatLeftLabel)
        end
    else
        return self.Label
    end
end 

function TabLeftItem:Enabled(enabled)
    if enabled ~= nil then
        self._enabled = enabled
        if not self._Enabled then
            self._formatLeftLabel = ReplaceRstarColorsWith(self._formatLeftLabel, "~c~")
        else
            self:Label(self._label)
        end
        if self.Parent ~= nil and self.Parent.Base.Parent ~= nil and self.Parent.Base.Parent:Visible() then
            local tab = IndexOf(self.Parent.Base.Parent.Tabs, self.Parent) - 1
            local leftItem = IndexOf(self.Parent.LeftItemList, self) - 1
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("ENABLE_LEFT_ITEM", tab, leftItem, self._enabled)
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("UPDATE_LEFT_ITEM_LABEL", tab, leftItem, self._formatLeftLabel)
        end
    else
        return self._enabled
    end
end

function TabLeftItem:Hovered(hover)
    if hover ~= nil then
        self._hovered = hover
    else
        return self._hovered
    end
end

function TabLeftItem:Selected(selected)
    if selected ~= nil then
        self._selected = selected
        if self._selected then
            self._formatLeftLabel = self._formatLeftLabel:gsub("~w~", "~l~")
            self._formatLeftLabel = self._formatLeftLabel:gsub("~s~", "~l~")
        else
            self._formatLeftLabel = self._formatLeftLabel:gsub("~l~", "~s~")
        end
        if self.Parent ~= nil and self.Parent.Base.Parent ~= nil and self.Parent.Base.Parent:Visible() then
            local tab = IndexOf(self.Parent.Base.Parent.Tabs, self.Parent) - 1
            local leftItem = IndexOf(self.Parent.LeftItemList, self) - 1
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("UPDATE_LEFT_ITEM_LABEL", tab, leftItem, self._formatLeftLabel)
        end
    else
        return self._selected
    end
end

function TabLeftItem:RightTitle(rtit)
    if rtit ~= nil then
        self.RightTitle = rtit
        if self.Parent ~= nil and self.Parent.Base.Parent ~= nil and self.Parent.Base.Parent:Visible() then
            local tab = IndexOf(self.Parent.Base.Parent.Tabs, self.Parent) - 1
            local leftItem = IndexOf(self.Parent.LeftItemList, self) - 1
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("ADD_RIGHT_TITLE", tab, leftItem, self.RightTitle, self.keymapRightLabel_1, self.KeymapRightLabel_2)
        end
    else
        return self.RightLabel
    end
end

function TabLeftItem:KeymapRightLabel_1(rtit)
    if rtit ~= nil then
        self.KeymapRightLabel_1 = rtit
        if self.Parent ~= nil and self.Parent.Base.Parent ~= nil and self.Parent.Base.Parent:Visible() then
            local tab = IndexOf(self.Parent.Base.Parent.Tabs, self.Parent) - 1
            local leftItem = IndexOf(self.Parent.LeftItemList, self) - 1
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("ADD_RIGHT_TITLE", tab, leftItem, self.RightTitle, self.keymapRightLabel_1, self.KeymapRightLabel_2)
        end
    else
        return self.KeymapRightLabel_1
    end
end

function TabLeftItem:KeymapRightLabel_2(rtit)
    if rtit ~= nil then
        self.KeymapRightLabel_2 = rtit
        if self.Parent ~= nil and self.Parent.Base.Parent ~= nil and self.Parent.Base.Parent:Visible() then
            local tab = IndexOf(self.Parent.Base.Parent.Tabs, self.Parent) - 1
            local leftItem = IndexOf(self.Parent.LeftItemList, self) - 1
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("ADD_RIGHT_TITLE", tab, leftItem, self.RightTitle, self.keymapRightLabel_1, self.KeymapRightLabel_2)
        end
    else
        return self.KeymapRightLabel_2
    end
end

function TabLeftItem:UpdateBackground(txd, txn, resizeType)
    self.TextureDict = txd
    self.TextureName = txn
    self.LeftItemBGType = resizeType
    if self.Parent ~= nil and self.Parent.Base.Parent ~= nil and self.Parent.Base.Parent:Visible() then
        local tab = IndexOf(self.Parent.Base.Parent.Tabs, self.Parent) - 1
        local leftItem = IndexOf(self.Parent.LeftItemList, self) - 1
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("UPDATE_LEFT_ITEM_RIGHT_BACKGROUND", tab, leftItem, txd, txn, resizeType)
    end
end

File: ScaleformUI_Csharp/Menus/Pause Menus/PauseMenu/Tabs/TabLeftItem.cs
using ScaleformUI.Elements;
using ScaleformUI.Menu;

namespace ScaleformUI.PauseMenu
{
    public enum LeftItemType
    {
        Empty,
        Info,
        Statistics,
        Settings,
        Keymap
    }

    public enum LeftItemBGType
    {
        Full,
        Masked,
        Resized
    }

    public delegate void IndexChangeEvent(SettingsItem item, int index);
    public delegate void ActivatedEvent(TabLeftItem item, int index);

    public class TabLeftItem : PauseMenuItem
    {
        internal UIMenuItem _internalItem {  get; set; }
        private bool enabled = true;
        private string label;
        internal string _formatLeftLabel;
        private SColor mainColor = SColor.HUD_Pause_bg;
        private SColor highlightColor = SColor.HUD_White;
        private string textTitle;
        private string keymapLeftLabel;
        private string keymapRightLabel_1;
        private string keymapRightLabel_2;
        public string TextureDict { get; private set; }
        public string TextureName { get; private set; }
        public LeftItemBGType LeftItemBGType { get; private set; }

        internal ItemFont _labelFont = ScaleformFonts.CHALET_LONDON_NINETEENSIXTY;
        internal ItemFont _rightLabelFont = ScaleformFonts.CHALET_LONDON_NINETEENSIXTY;
        public LeftItemType ItemType { get; internal set; }
        public string Label
        {
            get => label;
            set
            {
                label = value;
                _formatLeftLabel = value.StartsWith("~") ? value : "~s~" + value;
                if (Selected)
                {
                    _formatLeftLabel = _formatLeftLabel.Replace("~w~", "~l~");
                    _formatLeftLabel = _formatLeftLabel.Replace("~s~", "~l~");
                }
                else
                {
                    _formatLeftLabel = _formatLeftLabel.Replace("~l~", "~s~");
                }
                if (Parent != null && Parent.Visible)
                {
                    int it = Parent.LeftItemList.IndexOf(this);
                    Parent.Parent._pause._pause.CallFunction("UPDATE_LEFT_ITEM_LABEL", it, _formatLeftLabel);
                }
            }
        }
        public SColor MainColor { get => mainColor; set => mainColor = value; }
        public SColor HighlightColor { get => highlightColor; set => highlightColor = value; }
        public bool Enabled
        {
            get => enabled;
            set
            {
                enabled = value;
                if (!value)
                    _formatLeftLabel = _formatLeftLabel.ReplaceRstarColorsWith("~c~");
                else
                    Label = label;

                if (Parent != null && Parent.Visible)
                {
                    int it = Parent.LeftItemList.IndexOf(this);
                    Parent.Parent._pause._pause.CallFunction("ENABLE_LEFT_ITEM", it, enabled);
                    Parent.Parent._pause._pause.CallFunction("UPDATE_LEFT_ITEM_LABEL", it, _formatLeftLabel);
                }
            }
        }

        public bool Hovered { get; internal set; }

        public bool Selected
        {
            get => selected;
            internal set
            {
                selected = value;

                if (value)
                {
                    _formatLeftLabel = _formatLeftLabel.Replace("~w~", "~l~");
                    _formatLeftLabel = _formatLeftLabel.Replace("~s~", "~l~");
                }
                else
                {
                    _formatLeftLabel = _formatLeftLabel.Replace("~l~", "~s~");
                }
                if (Parent != null && Parent.Visible)
                {
                    int it = Parent.LeftItemList.IndexOf(this);
                    Parent.Parent._pause._pause.CallFunction("UPDATE_LEFT_ITEM_LABEL", it, _formatLeftLabel);
                }
            }
        }

        public int ItemIndex { get; internal set; }
        public List<PauseMenuItem> ItemList = new List<PauseMenuItem>();
        private bool selected;

        public string RightTitle
        {
            get => textTitle;
            set
            {
                textTitle = value;
                if (Parent != null && Parent.Visible)
                {
                    int it = Parent.LeftItemList.IndexOf(this);
                    Parent.Parent._pause._pause.CallFunction("ADD_RIGHT_TITLE", it, textTitle, KeymapRightLabel_1, KeymapRightLabel_2);
                }
            }
        }

        public string KeymapRightLabel_1
        {
            get => keymapRightLabel_1;
            set
            {
                keymapRightLabel_1 = value;
                if (Parent != null && Parent.Visible)
                {
                    int it = Parent.LeftItemList.IndexOf(this);
                    Parent.Parent._pause._pause.CallFunction("ADD_RIGHT_TITLE", it, textTitle, keymapRightLabel_1, KeymapRightLabel_2);
                }
            }
        }
        public string KeymapRightLabel_2
        {
            get => keymapRightLabel_2;
            set
            {
                keymapRightLabel_2 = value;
                if (Parent != null && Parent.Visible)
                {
                    int it = Parent.LeftItemList.IndexOf(this);
                    Parent.Parent._pause._pause.CallFunction("ADD_RIGHT_TITLE", it, textTitle, KeymapRightLabel_1, keymapRightLabel_2);
                }
            }
        }

        public event IndexChangeEvent OnIndexChanged;
        public event ActivatedEvent OnActivated;
        public BaseTab Parent { get; set; }

        public TabLeftItem(string label, LeftItemType type) : this(label, type, SColor.HUD_Pause_bg, SColor.HUD_White) { }
        public TabLeftItem(string label, LeftItemType type, SColor mainColor, SColor highlightColor) : base(label)
        {
            Label = label;
            ItemType = type;
            MainColor = mainColor;
            HighlightColor = highlightColor;
            _internalItem = new UIMenuItem(label, "", SColor.HUD_Pause_bg, SColor.HUD_White);
        }
        public TabLeftItem(string label, LeftItemType type, ItemFont labelFont, SColor mainColor, SColor highlightColor) : base(label)
        {
            Label = label;
            ItemType = type;
            MainColor = mainColor;
            HighlightColor = highlightColor;
            _labelFont = labelFont;
            _internalItem = new UIMenuItem(label, "", SColor.HUD_Pause_bg, SColor.HUD_White);
        }

        /// <summary>
        /// Image suggested not bigger than 720 x 576
        /// </summary>
        /// <param name="txd"></param>
        /// <param name="txn"></param>
        public void UpdateBackground(string txd, string txn, LeftItemBGType resizeType)
        {
            TextureDict = txd;
            TextureName = txn;
            LeftItemBGType = resizeType;
            if (Parent != null && Parent.Visible && Parent.Parent != null && Parent.Parent.Visible && this.ItemType != LeftItemType.Empty)
            {
                Parent.Parent._pause._pause.CallFunction("UPDATE_LEFT_ITEM_RIGHT_BACKGROUND", Parent.Parent.Tabs.IndexOf(Parent), Parent.LeftItemList.IndexOf(this), txd, txn, (int)resizeType);
            }
        }

        public void AddItem(PauseMenuItem item)
        {
            item.Parent = this;
            ItemList.Add(item);
        }

        internal void IndexChanged()
        {
            OnIndexChanged?.Invoke(ItemList[ItemIndex] as SettingsItem, ItemIndex);
        }

        internal void Activated()
        {
            OnActivated?.Invoke(this, Parent.LeftItemList.IndexOf(this));
        }
    }
}
